timeline/src/Data/Timeline/Internal.hs

```diff
@@ -0,0 +1,299 @@
+        ++ ["----------Timeline--End---------------"]
+    where
+      showOneChange :: (UTCTime, a) -> String
+      showOneChange (t, x) = "since " <> show t <> ": " <> show x
+
+peek :: Timeline a -> UTCTime -> a
+peek Timeline {..} time = maybe initialValue snd $ Map.lookupLE time values
+
+data TimeRange = TimeRange
+  { trFrom :: Maybe UTCTime,
+    trTo :: Maybe UTCTime
+  }
+  deriving stock (Show, Eq, Ord, Generic)
+
+isTimeAfterRange :: UTCTime -> TimeRange -> Bool
+isTimeAfterRange t TimeRange {trTo} = maybe False (t >=) trTo
+
+instance FunctorWithIndex TimeRange Timeline where
+  imap :: (TimeRange -> a -> b) -> Timeline a -> Timeline b
+  imap f Timeline {..} =
+    Timeline
+      { initialValue = f initialRange initialValue,
+        values = flip Map.mapWithKey values $ \from value ->
+          let timeRange = TimeRange (Just from) (fst <$> Map.lookupGT from values)
+           in f timeRange value
+      }
+    where
+      initialRange = TimeRange Nothing $ fst <$> Map.lookupMin values
+
+instance FoldableWithIndex TimeRange Timeline
+
+instance TraversableWithIndex TimeRange Timeline where
+  itraverse :: (Applicative f) => (TimeRange -> a -> f b) -> Timeline a -> f (Timeline b)
+  itraverse f = sequenceA . imap f
+
+changes :: Timeline a -> Set UTCTime
+changes Timeline {values} = Map.keysSet values
+
+data Record a = Record
+  { effectiveFrom :: UTCTime,
+    effectiveTo :: Maybe UTCTime,
+    value :: a
+  }
+  deriving stock (Eq, Functor)
+
+type Getter s a = forall f. (Contravariant f) => (a -> f a) -> s -> f s
+
+recordEffectiveFrom :: Getter (Record a) UTCTime
+recordEffectiveFrom f r = contramap getEffectiveFrom (f (getEffectiveFrom r))
+  where
+    getEffectiveFrom :: Record a -> UTCTime
+    getEffectiveFrom Record {effectiveFrom} = effectiveFrom
+
+recordEffectiveTo :: Getter (Record a) (Maybe UTCTime)
+recordEffectiveTo f r = contramap getEffectiveTo (f (getEffectiveTo r))
+  where
+    getEffectiveTo :: Record a -> Maybe UTCTime
+    getEffectiveTo Record {effectiveTo} = effectiveTo
+
+recordValue :: Getter (Record a) a
+recordValue f r = contramap getValue (f (getValue r))
+  where
+    getValue :: Record a -> a
+    getValue Record {value} = value
+
+-- | 'makeRecord' returns 'Nothing' if @effectiveTo@ is not greater than @effectiveFrom@
+makeRecord ::
+  -- | effective from
+  UTCTime ->
+  -- | optional effective to
+  Maybe UTCTime ->
+  -- | value
+  a ->
+  Maybe (Record a)
+-- can't write a makeRecordTH because UTCTime has no Lift instance
+makeRecord effectiveFrom effectiveTo value =
+  if maybe False (effectiveFrom >=) effectiveTo
+    then Nothing
+    else Just Record {..}
+
+instance (TH.Lift a) => TH.Lift (Record a) where
+  liftTyped Record {..} =
+    [||
+    Record
+      (unLiftUTCTime $$(TH.liftTyped $ LiftUTCTime effectiveFrom))
+      (fmap unLiftUTCTime $$(TH.liftTyped $ LiftUTCTime <$> effectiveTo))
+      $$(TH.liftTyped value)
+    ||]
```

Comment:
Is this not derivable?

Reply:
Not derivable unless we use LiftUTCTime directly in `Record`, which is inconvenient.